rfd/0136-modern-signature-algorithms.md

```diff
@@ -0,0 +1,777 @@
+This will be the initial default, changing the default to `balanced-v1` will
+be a breaking change that will be made in either a major version release of
+Teleport, or with a version bump to the `teleport.yaml` config.
+
+The `balanced-v1` suite will use a modern set of algorithms selected to
+balance security, compatibility, and performance.
+The proposed selection for this suite is to use Ed25519 as the default choice,
+and ECDSA with the NIST P-256 curve in cases where Teleport is likely to
+interact with third-party software that does not support Ed25519.
+The exception is the database CA, where it looks like it will be necessary to
+continue using RSA in order to continue supporting our current Snowflake
+integration.
+
+The `fips-v1` suite will only use key types and signature algorithms that are
+approved by FIPS 186-5 *and* are supported in Go's `GOEXPERIMENT=boringcrypto`
+mode.
+FIPS 186-5 actually approves all algorithms in the `balanced-v1` suite (and
+even the `legacy` suite), but Go's BoringCrypto integration does NOT support
+Ed25519.
+This suite will be based off of the `balanced-v1`, but all uses of Ed25519 will
+be replaced by ECDSA with the NIST P-256 curve.
+Teleport will fail to start in FIPS mode unless the `fips-v1` or `legacy` suite
+is selected.
+The `fips-v1` algorithm suite will be available without FIPS mode, to enable
+migration to FIPS mode.
+
+The `hsm-v1` suite will be based off of the `balanced-v1` suite, but all uses of
+Ed25519 *for CA keys only* will be replaced by ECDSA with the NIST P-256 curve.
+This is necessary and sufficient for all of the HSMs and KMSs we currently
+support and test for (YubiHSM2, AWS CloudHSM, and GCP KMS).
+This is very similar to the `fips-v1` suite, but "subjects" (users, hosts, etc)
+will still be allowed to use Ed25519 keys, it is only the CAs that are limited
+to the algorithms supported by the HSM.
+Whenever the default suite is updated from "legacy" to "balanced-v1", it will be
+overridden by `fips-v1` if and HSM or KMS is configured.
+Clusters with HSMs that must use FIPS algorithms should used the `fips-v1`
+suite.
+
+### Configuration
+
+The key types and signature algorithms used cluster-wide will be configurable
+via `cluster_auth_preference` and `teleport.yaml`
+
+We want it configurable via `teleport.yaml` so that you can start a new cluster
+and the CA keys will be automatically generated at first start with the correct
+algorithms, so you don't have to immediately edit the `cap` and then rotate all
+of your brand-new CAs.
+
+We want it configurable via `cluster_auth_preference` as well so that it can be
+configurable for Cloud users.
+
+If a value is set in the `cluster_auth_preference`, it will completely override
+the setting from `teleport.yaml`
+
+```yaml
+# teleport.yaml
+version: v3
+teleport:
+  auth_service:
+    enabled: true
+
+    # supported values are balanced-v1, fips-v1, and legacy
+    signature_algorithm_suite: balanced-v1
```

Comment:
in `v4` this should be the default starting `v14`. All existing deployments will be in `v3` config version using `legacy` by default.